Batteries/Data/Fin/Basic.lean

```diff
@@ -17,3 +17,71 @@ alias enum := Array.finRange
 
 @[deprecated (since := "2024-11-15")]
 alias list := List.finRange
+
+/-- Dependent version of `Fin.foldlM`. -/
+@[specialize] def dfoldlM [Monad m] (n : Nat) (α : Fin (n + 1) → Sort _)
+    (f : ∀ (i : Fin n), α i.castSucc → m (α i.succ)) (init : α 0) : m (α (last n)) :=
+  loop 0 (Nat.zero_lt_succ n) init where
+  /-- Inner loop for `Fin.dfoldlM`.
+    ```
+  Fin.foldM.loop n α f i h xᵢ = do
+    let xᵢ₊₁ ← f i xᵢ
+    ...
+    let xₙ ← f (n-1) xₙ₋₁
+    pure xₙ
+  ```
+  -/
+  loop (i : Nat) (h : i < n + 1) (x : α ⟨i, h⟩) : m (α (last n)) :=
+    if h' : i < n then
+      (f ⟨i, h'⟩ x) >>= loop (i + 1) (Nat.succ_lt_succ h')
+    else
+      haveI : ⟨i, h⟩ = last n := by ext; simp; omega
+      _root_.cast (congrArg (fun i => m (α i)) this) (pure x)
+
+/-- Dependent version of `Fin.foldrM`. -/
+@[specialize] def dfoldrM [Monad m] (n : Nat) (α : Fin (n + 1) → Sort _)
+    (f : ∀ (i : Fin n), α i.succ → m (α i.castSucc)) (init : α (last n)) : m (α 0) :=
+  loop n (Nat.lt_succ_self n) init where
+  /-- Inner loop for `Fin.foldRevM`.
+    ```
+  Fin.foldRevM.loop n α f i h xᵢ = do
+    let xᵢ₋₁ ← f (i+1) xᵢ
+    ...
+    let x₁ ← f 1 x₂
+    let x₀ ← f 0 x₁
+    pure x₀
+  ```
+  -/
+  loop (i : Nat) (h : i < n + 1) (x : α ⟨i, h⟩) : m (α 0) :=
+    if h' : i > 0 then
+      (f ⟨i - 1, by omega⟩ (by simpa [Nat.sub_one_add_one_eq_of_pos h'] using x))
+        >>= loop (i - 1) (by omega)
+    else
+      haveI : ⟨i, h⟩ = 0 := by ext; simp; omega
+      _root_.cast (congrArg (fun i => m (α i)) this) (pure x)
+
+/-- Dependent version of `Fin.foldl`. -/
+@[specialize] def dfoldl (n : Nat) (α : Fin (n + 1) → Sort _)
+    (f : ∀ (i : Fin n), α i.castSucc → α i.succ) (init : α 0) : α (last n) :=
+  loop 0 (Nat.zero_lt_succ n) init where
+  /-- Inner loop for `Fin.dfoldl`. `Fin.dfoldl.loop n α f i h x = f n (f (n-1) (... (f i x)))` -/
+  loop (i : Nat) (h : i < n + 1) (x : α ⟨i, h⟩) : α (last n) :=
```

Comment:
Sorry, my previous suggestion had a mistake... `specialize` goes on the inner loop. Same for the others.
 
```suggestion
def dfoldl (n : Nat) (α : Fin (n + 1) → Sort _)
    (f : ∀ (i : Fin n), α i.castSucc → α i.succ) (init : α 0) : α (last n) :=
  loop 0 (Nat.zero_lt_succ n) init where
  /-- Inner loop for `Fin.dfoldl`. `Fin.dfoldl.loop n α f i h x = f n (f (n-1) (... (f i x)))` -/
  @[specialize] loop (i : Nat) (h : i < n + 1) (x : α ⟨i, h⟩) : α (last n) :=
```

`specialize` makes lean generate new code for each `f`. It's similar to inlining `f` rather than the other way around.

Reply:
I think the inner loop should also be `semireducible`. There was a recent issue where it was a problem that it is `irreducible` because of the well-founded recursion.

Reply:
That would have been an issue for the non-dependent versions as well, right? I don't see `semireducible` attribute for `Fin.foldl.loop`, say.

Reply:
https://github.com/leanprover/lean4/blob/d9d54c1f99363f42d1485f6b67f281a5985f684e/src/Init/Data/Fin/Fold.lean#L16C1-L17C1

Reply:
Thanks, I was looking at an older version. Will make changes

Reply:
Actually, it's weird that only `Fin.foldl.loop` has `semireducible` attribute, but not other variants. I also added `semireducible` to only `Fin.dfoldl.loop`.

Reply:
The better alternative would be to use `match` instead of `if ... then ... else ...` so that structural recursion can be used instead.